sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -425,7 +425,8 @@ class Analyzer(
              */
             j
           case Some((oldRelation, newRelation)) =>
-            val attributeRewrites = AttributeMap(oldRelation.output.zip(newRelation.output))
+            val attributeRewrites =
+              AttributeMap(oldRelation.output.zip(newRelation.output).filter(x => x._1 != x._2))
```

Comment:
It can avoid extra compare and replacement in the subsequent transformation. For example, in the test case I added, this change can avoid making a copy for the underlying `LogicalRDD`. 

That is my original fix. If you think this change is not needed, I can easily remove it.